File: src/main/kotlin/io/github/ephemient/aoc2019/Day14.kt
package io.github.ephemient.aoc2019

class Day14(lines: List<String>) {
    private val rules: Map<String, Pair<Long, Map<String, Long>>> = lines.associate { line ->
        val items = PATTERN.findAll(line).map { match ->
            val (n, id) = match.destructured
            id to n.toLong()
        }.toMutableList()
        val (id, n) = items.removeAt(items.lastIndex)
        id to (n to items.toMap())
    }

    private fun produce(quantity: Long): Long {
        val mats = mutableMapOf("FUEL" to quantity)
        while (true) {
            val (key, n) = mats.entries.find { (key, n) -> key != "ORE" && n > 0 } ?: break
            val (m, srcs) = checkNotNull(rules[key]) { "null key=$key" }
            val x = (n + m - 1) / m
            mats[key] = n - m * x
            for ((k, v) in srcs) {
                mats[k] = mats.getOrElse(k) { 0 } + x * v
            }
        }
        return mats.getOrElse("ORE") { 0 }
    }

    fun part1(): Long = produce(1)

    fun part2(): Long {
        var good = MAX / produce(1)
        var bad: Long? = null
        while (bad == null || good < bad - 1) {
            val quantity = bad?.let { (good + it) / 2 } ?: good * 2
            val ore = produce(quantity)
            if (ore < MAX) {
                good = quantity
            } else if (ore == MAX) {
                return quantity
            } else {
                bad = quantity
            }
        }
        return good
    }

    companion object {
        private val PATTERN =
            """(\d+) (\w+)""".toRegex()
        private const val MAX = 1_000_000_000_000
    }
}

File: src/day14.rs
use super::util;
use regex::Regex;
use std::cmp::Ordering;
use std::collections::HashMap;
use std::error;
use std::ops::{Add, Div, Mul, Sub};
use std::str::FromStr;

type Rules<T> = HashMap<String, (T, HashMap<String, T>)>;

fn produce<T>(rules: &Rules<T>, quantity: T) -> T
where
    T: Clone
        + Default
        + Ord
        + From<i32>
        + Add<Output = T>
        + Sub<Output = T>
        + Mul<Output = T>
        + Div<Output = T>,
{
    let ore = "ORE".to_string();
    let mut mats = HashMap::new();
    mats.insert("FUEL".to_string(), quantity);
    while let Some((key, n)) = mats
        .iter()
        .filter_map(|(key, n)| {
            if *key != ore && *n > T::default() {
                Some((key.clone(), n.clone()))
            } else {
                None
            }
        })
        .nth(0)
    {
        let (m, srcs) = &rules[&key];
        let x = (n.clone() + m.clone() - T::from(1)) / m.clone();
        mats.insert(key, n - m.clone() * x.clone());
        for (k, v) in srcs.iter() {
            let y = mats.entry(k.to_string()).or_default();
            *y = y.clone() + v.clone() * x.clone();
        }
    }
    mats.get(&ore).cloned().unwrap_or_default()
}

fn parse<'a, I, S, T>(lines: I) -> Result<Rules<T>, Box<dyn error::Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
    T: Clone + FromStr,
    <T as FromStr>::Err: error::Error + Send + Sync + 'static,
{
    lazy_static! {
        static ref RE: Regex = Regex::new(r#"(\d+) (\w+)"#).unwrap();
    }
    lines
        .into_iter()
        .map(|line| {
            let mut items = RE
                .captures_iter(line.as_ref())
                .map(|cap| -> Result<_, <T as FromStr>::Err> {
                    Ok((cap[2].to_string(), cap[1].parse::<T>()?))
                })
                .collect::<Result<Vec<_>, _>>()?;
            let (key, n) = items.pop().ok_or(util::Error)?;
            Ok((key, (n, items.iter().cloned().collect())))
        })
        .collect()
}

pub fn part1<'a, I, S>(lines: I) -> Result<i64, Box<dyn error::Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    Ok(produce(&parse::<_, _, i64>(lines)?, 1))
}

pub fn part2<'a, I, S>(lines: I) -> Result<i64, Box<dyn error::Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let rules = parse::<_, _, i64>(lines)?;
    let target = 1_000_000_000_000i64;
    let mut good = target / produce(&rules, 1);
    let mut bad: Option<i64> = None;
    while bad.map_or(true, |bad| good < bad - 1) {
        let mid = bad.map_or_else(|| good * 2, |bad| (good + bad) / 2);
        match produce(&rules, mid).cmp(&target) {
            Ordering::Less => good = mid,
            Ordering::Equal => return Ok(mid),
            Ordering::Greater => bad = Some(mid),
        }
    }
    Ok(good)
}
